### DISCUSSION TITLE
Does Milvus support batch import directly by importing data into the message queue?

### DISCUSSION THREAD

**Discussion by xuefeimiaoao** [milvus-io/milvus#26714] [Q&A and General discussion] [answered]
Does Milvus support batch import directly by importing data into the message queue, such as kafka?

**Comment by yhmo** [comment]
Do you mean insert vectors batch by batch?

**Reply by xiaofan-luan** [reply]
check bulk insert feature https://milvus.io/docs/bulk_insert.md

**Reply by xuefeimiaoao** [reply]
In my case, thousands of `pymilvus.orm.utility.do_bulk_insert` may executed concurrently with column-based bulk insert. Is there any strategy to enhance the efficiency of my batch import process? Could kafka be exposed to users?

**Reply by yhmo** [reply, marked as answer]
bulkinsert doesn't pass the data to pulsar/kafka. The data files are directly parsed by data node and converted to segments.
Each time you call do_bulk_insert(), a bulkinsert task is generated and send to a data node to execute. Each data node can only execute bulkinsert tasks one by one. 
There is a task list maintained in rootcoord. If you concurrently call do_bulk_insert(), the tasks will be accumulated in rootcoord, waiting for datanode to consume. concurrently call do_bulk_insert() doesn't help to improve data ingestion.
If there are 2 or more data nodes in the cluster, they can execute bulkinsert tasks parallelly.

**Reply by xuefeimiaoao** [reply]
Thand you!
